### ISSUE TITLE
Support configuring per-method retryPolicy for gRPC channel

### ISSUE THREAD

**Issue by vladykin** [quarkusio/quarkus#16241]
## Description

`io.quarkus.grpc.runtime.supports.Channels` is responsible for producing instances of `io.grpc.Channel`.
Most of `io.grpc.Channel` properties are set from application.properties. But not all.
I'm looking for a way to call `io.grpc.ManagedChannelBuilder#defaultServiceConfig` to configure retry policy. This is needed to make gRPC retry actually work.

Setting `quarkus.grpc.clients."service-name".retry=true` is not enough. It enables retry subsystem, but retry policy must be enabled and configured for each method individually.

See this question and its links for more details: https://stackoverflow.com/questions/65701091/how-to-get-grpcs-retry-mechanism-to-work-using-grpc-java-in-kubernetes-cluster

Also official docs for gRPC retry: https://github.com/grpc/proposal/blob/master/A6-client-retries.md

**Comment by cescoffier** [comment]
Did you try with the @Retry annotation?

At the moment we don't support the retry gRPC extension, but that's something we can add.

**Comment by vladykin** [comment]
No, I didn't try this annotation. Do you mean that I should put it on a method in my app that calls gRPC stub methods? And as result my method will be called again if it throws exception?
Well, that sounds like a viable workaround for blocking methods. But I suspect it won't work for methods returning CompletionStage and Uni/Multi, where nothing is thrown from method in case of failure.

For Uni and Multi I can use their built-in retry capabilities. That's what I'm going to do if native gRPC retry is not available.

BTW, what's the point of exposing retry-related properties (`quarkus.grpc.clients."service-name".retry`, etc) if gRPC retry doesn't work regardless of the configured values of these properties?

**Comment by cescoffier** [comment]
Good question about the configuration. @michalszynkiewicz do we have some kind of support? 

About @Retry in the next version it will work with method returning Uni. I don't believe it will work for Multi. I said next version, as it needs Smallrye Fault Tolerance 5.x which will come with a Quarkus 2.0

**Comment by michalszynkiewicz** [comment]
`@Retry` works for methods that return `CompletionStage` but you need to also add `@Asynchronous` annotation.

**Comment by michalszynkiewicz** [comment]
I will take a look at it when I have a while.

**Comment by mkouba** [comment]
There are couple of things we need to consider:

- The retry policy is part of the service config which is bound to a `Channel` instance via `io.grpc.internal.AbstractManagedChannelImplBuilder.defaultServiceConfig()`
- A `Channel` instance is created per client configured via `GrpcClientConfiguration`
- So in therory, a `Channel` instance can be shared for multiple services...

The tricky part is that if we want to configure the retry policy per method then the config properties would have include the method name, e.g. something like `quarkus.grpc.clients.hello.method.SayHello.retry.max-attempts=2`. But (1) properties names are very long and (2) even then we would not be able to distinguish methods of a specific service.

We can also support per-client configuration that would be applied to all methods of all services:
```
quarkus.grpc.clients.hello.retry.max-attempts=2
quarkus.grpc.clients.hello.retry.initial-backoff=0.5s
quarkus.grpc.clients.hello.retry.max-backoff=1s
quarkus.grpc.clients.hello.retry.backoff-multiplier=1
quarkus.grpc.clients.hello.retry.retryable-status-codes=UNAVAILABLE
```

**Comment by ivansenic** [comment]
What's the state of this issue, we are hitting a `UNAVAILABLE: Abrupt GOAWAY closed sent stream.` when grpc calls are load balanced using the nginx. We enabled retries, but seems as @vladykin noted, without retrying policy it is not enough. Can we do something to push this issue forward, at least with the per-service configuration?

Is there any workaround currently to actually set the `RetryPolicy` manually, without using the configuration? 

If you need man power on pushing this forward, I would be happy to help and contribute.

**Comment by mkouba** [comment]
CC @alesj

**Comment by alesj** [comment]
@ivansenic @mkouba what would then be the ideal solution for this?

Atm I see this 

```
                        .maxRetryAttempts(config.maxRetryAttempts)
            }

            if (config.retry) {
                builder.enableRetry();

```

**Comment by cescoffier** [comment]
Wondering if we should just use the @Retry from fault tolerance?

**Comment by mkouba** [comment]
> Wondering if we should just use the @Retry from fault tolerance?

`@Retry` would not work with `Multi`, or? At least you mentioned it in [one of the previous comments](https://github.com/quarkusio/quarkus/issues/16241#issuecomment-813576312).

**Comment by mkouba** [comment]
> Atm I see this

@alesj where do you see this snippet?

**Comment by cescoffier** [comment]
ah yes, won't work for multi, it would retry the full multi

**Comment by alesj** [comment]
@mkouba that's part of gRPC's Channels code ... 
* https://github.com/quarkusio/quarkus/blob/main/extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/Channels.java#L171

**Comment by mkouba** [comment]
> @mkouba that's part of gRPC's Channels code ...
> 
>     * https://github.com/quarkusio/quarkus/blob/main/extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/Channels.java#L171

Yes, so this merely enables the retry subsystem per channel but we need a way to configure the retry policy per method.

**Comment by ivansenic** [comment]
The simplest solution that I can think of is to enable the specification of the service config as JSON, like:

```
quarkus.grpc.clients.my-client.default-service-config: |
 {
   .. full service config here
 }
```

But I am not sure if parsing this would work? We eventually just need a `Map<String,Object>` to be passed to a `io.grpc.internal.AbstractManagedChannelImplBuilder#defaultServiceConfig`.

Or eventually a full-blow configuration support for defining service config using props and then creating that map input from the config.

**Comment by mkouba** [comment]
> Or eventually a full-blow configuration support for defining service config using props and then creating that map input from the config.

Yes, I think that this was the original plan. But as I mentioned in https://github.com/quarkusio/quarkus/issues/16241#issuecomment-863223246 the tricky part is how to define the format used, e.g. `quarkus.grpc.clients.hello.method.SayHello.retry.max-attempts=2` isn't exactly convenient...

**Comment by ivansenic** [comment]
OK, what about the `ManagedChannelBuilderCustomizer`, where we could define one in the app and that would receive the builder so we can actually manually setup whatever we want? And all customizers could be processed here after the interceptors: https://github.com/quarkusio/quarkus/blob/340b85c1b8b005750d6009a1990f59a77f32bff5/extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/Channels.java#L213-L218

**Comment by mkouba** [comment]
> OK, what about the `ManagedChannelBuilderCustomizer`, where we could define one in the app and that would receive the builder so we can actually manually setup whatever we want? And all customizers could be processed here after the interceptors:

Sounds reasonable, except that we would give the users a way to shoot themselves in the foot. What about `DefaultServiceConfigCustomizer` instead? The customizer would accept the client name and the relevant `GrpcClientConfiguration` and then returns the `Map<String, Object>` config.

**Comment by cescoffier** [comment]
BTW, just to explain why I don't like the "properties" approach: it creates ambiguity when using env variables (as everything will be in uppercase).

The customizer approach should be qualified as: for every advanced users only. You can quickly end up with underperforming and incorrect channel configuration.

**Comment by ivansenic** [comment]
> > OK, what about the `ManagedChannelBuilderCustomizer`, where we could define one in the app and that would receive the builder so we can actually manually setup whatever we want? And all customizers could be processed here after the interceptors:
> 
> Sounds reasonable, except that we would give the users a way to shoot themselves in the foot. What about `DefaultServiceConfigCustomizer` instead? The customizer would accept the client name and the relevant `GrpcClientConfiguration` and then returns the `Map<String, Object>` config.

Works as well. Yes customizers should be for advanced users..

**Comment by alesj** [comment]
> The customizer would accept the client name and the relevant GrpcClientConfiguration and then returns the Map<String, Object> config.

What would we do then with this `Map<String, Object>` ?

Imho, that `ManagedChannelBuilderCustomizer` sounds a lot easier ... if you know what you're doing. :-)

**Comment by mkouba** [comment]
> What would we do then with this `Map<String, Object>` ?

Just use it as an argument for `AbstractManagedChannelImplBuilder.defaultServiceConfig(Map<String, ?>)`? We could also introduce our own API/builder so that it's clear what service config is supported.

> Imho, that ManagedChannelBuilderCustomizer sounds a lot easier ... if you know what you're doing. :-)

**Only** if you **really** know what you're doing :D

**Comment by alesj** [comment]
What if we just do it all-in-one ... combine the two ideas ... ?

```
ChannelBuilderCustomizer {
  Map<String, Object> customize(String name, GrpcClientConfiguration config, ChannelBuilder builder);
}
```

**Comment by alesj** [comment]
> I can provide the impl if needed..

It's not just the impl, it's also tests, docs, ...

**Comment by vikasmc** [comment]
Hi,
Any update on this feature? did anyone find a workaround for this?

**Comment by cescoffier** [comment]
I like @alesj proposal. The customizer approach gives the right level of flexibility.

**Comment by cescoffier** [comment]
@alesj do you plan to work on your proposal? It could be beneficial.

**Comment by alesj** [comment]
@cescoffier OK, let me move it up the list on my TODO stuff ...

**Comment by maperz22** [comment]
Hi! I know this is closed but it might be useful to let users know that setting `quarkus.grpc.clients.clientName.retry=true` is not enough to enable the retry mechanism. I had had to spend a couple of days investigating before I had found this thread. 
On this page: https://quarkus.io/guides/grpc-service-consumption#custom-channel-building it is not stated anywhere.

Just to save time for other people having this issue (OR NOT because they might even not know that their retry mechanism does not work) it would be great to update the info in the docs 🙏
